### ISSUE TITLE
Poll: Name a compiler option for including `undefined` when a property resolves to an index signature

### ISSUE THREAD

**Issue by RyanCavanaugh** [microsoft/TypeScript#40435]
We need a name for the option described in #13778

The exact behavior that this flag would enable is, briefly:
 * Read expressions of the form `expr[i]` will now be `T | undefined` where `T` is the index signature type
 * Read expressions of the form `expr.prop` will now be `T | undefined` where `T` is the index signature type if an index signature on `expr` exists but not a declared property named `prop`
 * `string` and `number` index signatures are both subjected to this treatment
 * Write expressions `expr[i] = v` and `expr.prop = v` are unaffected
 * Type expressions `TypeName[number]` and `TypeName["prop"]` are unaffected

Good candidate flag names should follow these rules and goals:
 * Do not start with `strict` since this flag would not get turned on by the global `--strict` setting
 * Follow an existing convention (see the list at [(https://aka.ms/tsconfig.json](https://aka.ms/tsconfig.json))
 * Convey the intent of the flag along with its behavior, somehow
 * Avoid problematic terms (e.g. do not reference mental illnesses)

# This is an *open-ended approval voting* issue. Rules:

  * Use 👍 reaction on as many as you like
    * Top-upvoted options will be given the most consideration
    * This is an advisory vote; final say is still up to the TypeScript team
  * You may propose new options; comments must be of exactly the form \`--proposedOptionName\`
    * Follow the above guidance - no `--strictFoo`!
  * No discussion here; discuss in #13778. Non-option comments will be deleted
  * Sharing this link on social media is encouraged!

I will seed this with some of the names suggested at our last meeting. We may also cull options that aren't receiving a lot of votes to help keep the comment count manageable.

**Comment by RyanCavanaugh** [comment]
`--pedanticIndexSignatures`

**Comment by RyanCavanaugh** [comment]
`--stringentIndexSignatures`

**Comment by RyanCavanaugh** [comment]
`--noUncheckedIndexedAccess`

**Comment by RyanCavanaugh** [comment]
`--undefinedIndexSignatures`

**Comment by RyanCavanaugh** [comment]
`--skepticalIndexAccesses`

**Comment by RyanCavanaugh** [comment]
`--noImplicitBounds`

**Comment by danvk** [comment]
`--paranoidIndexSignatures`

**Comment by johnnyreilly** [comment]
`--indexIsItThough`

**Comment by orta** [comment]
`--rigorousIndexAccess`

**Comment by danvk** [comment]
`--carefulIndexAccess`

**Comment by jack-williams** [comment]
```--partialIndexedAccess```

**Comment by awinograd** [comment]
`--maybeIndexAccess`

**Comment by OliverJAsh** [comment]
`--indexNullChecks`

**Comment by seajaysec** [comment]
`—indexSchmindex`

**Comment by johnnyreilly** [comment]
I realise it's not the point of this thread, but I'm super interested if it's a bot or a human deleting the comedy options... I'm tending towards thinking it is because there's definitely some comedy leaking through undetected by AI; `--pedanticIndexSignatures` do not go gentle into that good night

**Comment by kirillgroshkov** [comment]
`—safeIndexAccess`

**Comment by danvk** [comment]
`--soundIndexedAccess`

**Comment by thw0rted** [comment]
`--includeUndefinedInIndexSignatures`

**Comment by thw0rted** [comment]
`--includeUndefinedInIndexAccess`

**Comment by SimonFarrugia** [comment]
--orUndefinedIndexType

**Comment by alex-okrushko** [comment]
> --noUncheckedIndexedAccess

I think this one nails it, however I'd adjust it to `--noUncheckedIndexAccess` (uncheck-ed and index-ed it a feels a bit too much of -ed).

**Comment by osyrisrblx** [comment]
While this seems like a much safer option, it would be nice to opt-out of this behavior with a separate syntax for the rare occasion where you're 100% sure it's always defined.
```TS
interface X {
    [index: string]!: number;
}
```

Maybe `!:` could assert always defined?

**Comment by zeusdeux** [comment]
I'd say `--checkIndexedAccess` is better than the its double negated counterpart `--noUncheckedIndexedAccess`

**Comment by orta** [comment]
The double negation is to keep it consistent with [many others flags](https://www.typescriptlang.org/tsconfig) (noImplicitAny noImplicitThis noFallthroughCasesInSwitch noImplicitReturns noUnusedLocals noUnusedParameters)

**Comment by RyanCavanaugh** [comment]
We have a winner 🥇

**Comment by johnnyreilly** [comment]
Are you rolling with `--noUncheckedIndexedAccess` or @alex-okrushko's suggested tweak of `--noUncheckedIndexAccess`?
